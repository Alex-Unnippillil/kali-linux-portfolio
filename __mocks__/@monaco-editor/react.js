import React from 'react';
export default function MonacoMock({ value = '', onChange }) {
  return React.createElement('textarea', {
    value,
    onChange: e => onChange && onChange(e.target.value),
  });
}
