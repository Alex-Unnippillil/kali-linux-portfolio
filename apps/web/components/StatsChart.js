import React from 'react';

const StatsChart = ({ count, time }) => {
  const max = Math.max(count, time, 1);
  const cH = (count / max) * 80;
  const tH = (time / max) * 80;
  return (
    <svg viewBox="0 0 120 100" className="w-full h-24 mt-2">
      <rect x="10" y={90 - cH} width="40" height={cH} fill="#10b981" />
      <rect x="70" y={90 - tH} width="40" height={tH} fill="#3b82f6" />
      <text x="30" y="95" textAnchor="middle" fontSize="8" fill="white">
        candidates
      </text>
      <text x="90" y="95" textAnchor="middle" fontSize="8" fill="white">
        seconds
      </text>
    </svg>
  );
};

export default StatsChart;
