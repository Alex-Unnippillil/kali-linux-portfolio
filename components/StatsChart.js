import React, { useId, useMemo } from 'react';

const StatsChart = ({ count = 0, time = 0 }) => {
  const chartTitleId = useId();
  const chartDescId = useId();

  const maxValue = Math.max(count, time, 1);
  const chartHeight = 70;
  const baselineY = 90;
  const axisX = 40;
  const barWidth = 36;
  const barSpacing = 32;

  const numberFormatter = useMemo(
    () =>
      new Intl.NumberFormat(undefined, {
        maximumFractionDigits: 1,
      }),
    []
  );

  const bars = useMemo(
    () => [
      {
        key: 'candidates',
        label: 'Candidates processed',
        shortLabel: 'Candidates',
        value: count,
        unit: 'candidates',
        x: axisX + 10,
        color: 'var(--game-color-success)',
      },
      {
        key: 'seconds',
        label: 'Seconds elapsed',
        shortLabel: 'Seconds',
        value: time,
        unit: 'seconds',
        x: axisX + 10 + barWidth + barSpacing,
        color: 'var(--game-color-secondary)',
      },
    ],
    [axisX, barSpacing, barWidth, count, time]
  );

  const chartRight = bars[bars.length - 1].x + barWidth + 10;
  const axisColor = 'var(--color-ubt-grey, #aea79f)';
  const textColor = 'var(--color-text, #f5f5f5)';
  const transition =
    'height var(--motion-medium, 300ms) ease-out, y var(--motion-medium, 300ms) ease-out';

  const description = `Bar chart comparing ${numberFormatter.format(
    count
  )} candidates against ${numberFormatter.format(time)} seconds.`;

  return (
    <svg
      role="img"
      aria-labelledby={chartTitleId}
      aria-describedby={chartDescId}
      viewBox={`0 0 ${chartRight + 20} 120`}
      className="w-full h-28 mt-2"
    >
      <title id={chartTitleId}>Hash cracking progress overview</title>
      <desc id={chartDescId}>{description}</desc>
      <line
        x1={axisX}
        y1={baselineY - chartHeight}
        x2={axisX}
        y2={baselineY}
        stroke={axisColor}
        strokeWidth="1.5"
        vectorEffect="non-scaling-stroke"
      />
      <line
        x1={axisX}
        y1={baselineY}
        x2={chartRight}
        y2={baselineY}
        stroke={axisColor}
        strokeWidth="1.5"
        vectorEffect="non-scaling-stroke"
      />
      <g aria-hidden="true" fill={axisColor} fontSize="8">
        <text x={axisX - 4} y={baselineY + 6} textAnchor="end">
          0
        </text>
        <text x={axisX - 4} y={baselineY - chartHeight + 4} textAnchor="end">
          {numberFormatter.format(maxValue)}
        </text>
      </g>
      <text
        x={axisX / 2}
        y={baselineY - chartHeight / 2}
        textAnchor="middle"
        fontSize="9"
        fill={textColor}
        transform={`rotate(-90 ${axisX / 2} ${baselineY - chartHeight / 2})`}
      >
        Value
      </text>
      <text
        x={(axisX + chartRight) / 2}
        y={baselineY + 18}
        textAnchor="middle"
        fontSize="9"
        fill={textColor}
      >
        Metrics
      </text>
      {bars.map((bar) => {
        const height = (bar.value / maxValue) * chartHeight;
        const y = baselineY - height;
        return (
          <g key={bar.key}>
            <rect
              x={bar.x}
              y={y}
              width={barWidth}
              height={height}
              fill={bar.color}
              rx="4"
              style={{ transition }}
            >
              <title>
                {`${bar.label}: ${numberFormatter.format(bar.value)} ${bar.unit}`}
              </title>
            </rect>
            <text
              x={bar.x + barWidth / 2}
              y={baselineY + 10}
              textAnchor="middle"
              fontSize="8.5"
              fill={textColor}
            >
              {bar.shortLabel}
            </text>
          </g>
        );
      })}
    </svg>
  );
};

export default StatsChart;
