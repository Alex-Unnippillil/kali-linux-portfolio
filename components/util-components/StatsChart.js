import React, { useCallback, useEffect, useMemo, useRef, useState } from 'react';

const BASE_WIDTH = 320;
const BASE_HEIGHT = 140;
const CHART_HEIGHT = 90;
const PADDING_X = 24;
const PADDING_TOP = 12;
const BAR_WIDTH = 42;
const BAR_GAP = 28;
const TAP_THRESHOLD = 12; // px before we treat the gesture as a drag instead of a tap

const defaultFormatter = (value) =>
  typeof value === 'number' ? value.toLocaleString() : String(value ?? '');

const buildData = (count, time, data) => {
  if (Array.isArray(data) && data.length) {
    return data.map((entry, index) => ({
      key: entry.key ?? `value-${index}`,
      label: entry.label ?? entry.key ?? `Value ${index + 1}`,
      value: Number(entry.value) || 0,
      color: entry.color ?? '#3b82f6',
      formatter: entry.formatter ?? defaultFormatter,
    }));
  }

  return [
    {
      key: 'candidates',
      label: 'Candidates',
      value: Number(count) || 0,
      color: '#10b981',
      formatter: defaultFormatter,
    },
    {
      key: 'seconds',
      label: 'Seconds @1M/s',
      value: Number(time) || 0,
      color: '#3b82f6',
      formatter: defaultFormatter,
    },
  ];
};

const StatsChart = ({ count = 0, time = 0, data, ariaLabel = 'Candidate and runtime comparison' }) => {
  const preparedData = useMemo(() => buildData(count, time, data), [count, time, data]);
  const maxValue = useMemo(
    () => Math.max(1, ...preparedData.map((entry) => Math.max(0, entry.value))),
    [preparedData],
  );

  const bars = useMemo(() => {
    return preparedData.map((entry, index) => {
      const height = (Math.max(0, entry.value) / maxValue) * CHART_HEIGHT;
      const x = PADDING_X + index * (BAR_WIDTH + BAR_GAP);
      const y = PADDING_TOP + (CHART_HEIGHT - height);
      return {
        ...entry,
        x,
        y,
        height,
        width: BAR_WIDTH,
      };
    });
  }, [preparedData, maxValue]);

  const contentWidth = useMemo(() => {
    if (!bars.length) return BASE_WIDTH;
    return (
      PADDING_X * 2 +
      bars.length * BAR_WIDTH +
      (bars.length - 1) * BAR_GAP
    );
  }, [bars.length]);

  const containerRef = useRef(null);
  const pointerState = useRef(new Map());
  const [containerWidth, setContainerWidth] = useState(0);
  const [pan, setPan] = useState(0);
  const [tooltip, setTooltip] = useState(null);

  const minPan = contentWidth > BASE_WIDTH ? BASE_WIDTH - contentWidth : 0;
  const clampedPan = contentWidth <= BASE_WIDTH ? (BASE_WIDTH - contentWidth) / 2 : Math.min(0, Math.max(pan, minPan));

  useEffect(() => {
    if (!containerRef.current) return undefined;
    const node = containerRef.current;

    const updateWidth = () => {
      const rect = node.getBoundingClientRect();
      setContainerWidth(rect.width || BASE_WIDTH);
    };

    updateWidth();

    if (typeof ResizeObserver === 'undefined') {
      return undefined;
    }

    const resizeObserver = new ResizeObserver(updateWidth);
    resizeObserver.observe(node);

    return () => {
      resizeObserver.disconnect();
    };
  }, []);

  useEffect(() => {
    if (contentWidth <= BASE_WIDTH && pan !== 0) {
      setPan(0);
    }
  }, [contentWidth, pan]);

  const hideTooltip = useCallback(() => setTooltip(null), []);

  const showTooltipAt = useCallback(
    (index, clientX, clientY) => {
      if (!bars[index] || !containerRef.current) return;
      const rect = containerRef.current.getBoundingClientRect();
      const scaleXToPixels = rect.width ? rect.width / BASE_WIDTH : 1;
      const scaleYToPixels = rect.height ? rect.height / BASE_HEIGHT : 1;
      const barCenterX = bars[index].x + bars[index].width / 2 + clampedPan;
      const barTopY = bars[index].y;
      const left =
        clientX != null
          ? clientX - rect.left
          : barCenterX * scaleXToPixels;
      const top = clientY != null ? clientY - rect.top : barTopY * scaleYToPixels;
      setTooltip({
        index,
        left,
        top,
      });
    },
    [bars, clampedPan],
  );

  const scaleX = containerWidth ? BASE_WIDTH / containerWidth : 1;

  const updatePan = useCallback(
    (deltaPx) => {
      if (contentWidth <= BASE_WIDTH) return;
      setPan((previous) => {
        const next = previous + deltaPx * scaleX;
        if (next > 0) return 0;
        if (next < minPan) return minPan;
        return next;
      });
    },
    [contentWidth, minPan, scaleX],
  );

  const computeBarIndexFromPoint = useCallback(
    (clientX) => {
      if (!containerRef.current) return -1;
      const rect = containerRef.current.getBoundingClientRect();
      if (!rect.width) return -1;
      const xInChart = ((clientX - rect.left) / rect.width) * BASE_WIDTH - clampedPan;
      return bars.findIndex((bar) => xInChart >= bar.x && xInChart <= bar.x + bar.width);
    },
    [bars, clampedPan],
  );

  const handlePointerDown = useCallback(
    (event) => {
      pointerState.current.set(event.pointerId, {
        type: event.pointerType,
        x: event.clientX,
        y: event.clientY,
        prevX: event.clientX,
        prevY: event.clientY,
        downX: event.clientX,
        downY: event.clientY,
        moved: false,
      });
      if (event.pointerType === 'touch' && pointerState.current.size === 1) {
        // Hide tooltip when a new touch begins to avoid stale overlays
        hideTooltip();
      }
    },
    [hideTooltip],
  );

  const handlePointerMove = useCallback(
    (event) => {
      const stored = pointerState.current.get(event.pointerId);
      if (!stored) return;
      const next = {
        ...stored,
        prevX: stored.x,
        prevY: stored.y,
        x: event.clientX,
        y: event.clientY,
      };
      if (!stored.moved) {
        const delta = Math.hypot(next.x - next.downX, next.y - next.downY);
        if (delta > TAP_THRESHOLD) {
          next.moved = true;
        }
      }
      pointerState.current.set(event.pointerId, next);

      const touchPointers = Array.from(pointerState.current.values()).filter((pointer) => pointer.type === 'touch');
      if (touchPointers.length >= 2) {
        const avgDelta =
          touchPointers.reduce((sum, pointer) => sum + (pointer.x - pointer.prevX), 0) / touchPointers.length;
        if (!Number.isNaN(avgDelta)) {
          updatePan(avgDelta);
        }
      }
    },
    [updatePan],
  );

  const handlePointerEnd = useCallback(
    (event) => {
      const stored = pointerState.current.get(event.pointerId);
      pointerState.current.delete(event.pointerId);
      if (stored && event.pointerType === 'touch' && !stored.moved) {
        const index = computeBarIndexFromPoint(event.clientX);
        if (index !== -1) {
          showTooltipAt(index, event.clientX, event.clientY);
        }
      }

      if (pointerState.current.size === 0) {
        pointerState.current.clear();
      }
    },
    [computeBarIndexFromPoint, showTooltipAt],
  );

  const handleWheel = useCallback((event) => {
    // Prevent any chart-specific zoom behaviour but keep native page scroll alive.
    event.stopPropagation();
  }, []);

  const tooltipContent = useMemo(() => {
    if (!tooltip) return null;
    const bar = bars[tooltip.index];
    if (!bar) return null;
    const valueText = bar.formatter(bar.value);
    return (
      <div
        className="absolute bg-gray-900/95 text-xs text-white px-2 py-1 rounded shadow-lg pointer-events-none"
        style={{
          left: `${tooltip.left}px`,
          top: `${tooltip.top}px`,
          transform: 'translate(-50%, -100%)',
          maxWidth: '200px',
        }}
      >
        <div className="font-semibold">{bar.label}</div>
        <div>{valueText}</div>
      </div>
    );
  }, [bars, tooltip]);

  return (
    <div
      ref={containerRef}
      className="relative w-full h-32 mt-2 select-none"
      style={{ touchAction: 'pan-y' }}
      onPointerDown={handlePointerDown}
      onPointerMove={handlePointerMove}
      onPointerUp={handlePointerEnd}
      onPointerCancel={handlePointerEnd}
      onPointerLeave={(event) => {
        if (event.pointerType !== 'touch') {
          hideTooltip();
        }
      }}
      onWheel={handleWheel}
    >
      <svg viewBox={`0 0 ${BASE_WIDTH} ${BASE_HEIGHT}`} className="w-full h-full" role="img" aria-label={ariaLabel}>
        <desc>Interactive comparison chart optimised for touch gestures</desc>
        <g transform={`translate(${clampedPan}, 0)`}>
          {bars.map((bar, index) => (
            <g key={bar.key}>
              <rect
                x={bar.x}
                y={bar.y}
                width={bar.width}
                height={bar.height}
                fill={bar.color}
                rx="6"
                ry="6"
                className="transition-opacity duration-150 ease-out"
                onMouseEnter={(event) => showTooltipAt(index, event.clientX, event.clientY)}
                onMouseMove={(event) => showTooltipAt(index, event.clientX, event.clientY)}
                onMouseLeave={hideTooltip}
                onFocus={(event) => showTooltipAt(index, event.target.getBoundingClientRect().left + event.target.getBoundingClientRect().width / 2, event.target.getBoundingClientRect().top)}
                onBlur={hideTooltip}
                tabIndex={0}
                role="presentation"
              />
              <text
                x={bar.x + bar.width / 2}
                y={PADDING_TOP + CHART_HEIGHT + 18}
                textAnchor="middle"
                fontSize="10"
                fill="white"
              >
                {bar.label}
              </text>
            </g>
          ))}
        </g>
      </svg>
      {tooltipContent}
    </div>
  );
};

export default StatsChart;
