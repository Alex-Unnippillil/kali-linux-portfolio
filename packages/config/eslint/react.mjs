import globals from 'globals';
import { baseConfig, compat } from './base.mjs';

const reactConfig = [
  ...baseConfig,
  {
    languageOptions: {
      globals: globals.browser,
    },
    rules: {
      'no-top-level-window/no-top-level-window-or-document': 'off',
      'no-undef': 'off',
      'no-unused-vars': 'off',
      'react/react-in-jsx-scope': 'off',
    },
  },
  ...compat.config({
    extends: [
      'eslint:recommended',
      'plugin:@typescript-eslint/recommended',
      'plugin:react/recommended',
      'plugin:react-hooks/recommended',
    ],
    settings: {
      react: {
        version: 'detect',
      },
    },
  }),
  {
    rules: {
      '@typescript-eslint/no-unused-vars': ['error', { argsIgnorePattern: '^_' }],
      '@typescript-eslint/no-unused-expressions': ['error', { allowShortCircuit: true }],
      'react/react-in-jsx-scope': 'off',
    },
  },
];

export default reactConfig;
