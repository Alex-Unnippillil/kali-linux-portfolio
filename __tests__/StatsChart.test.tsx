import React from 'react';
import { render } from '@testing-library/react';

import StatsChart from '../components/StatsChart';

describe('StatsChart', () => {
  it('renders a snapshot for populated values', () => {
    const { container } = render(<StatsChart count={1200000} time={420} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders a snapshot when values are zero', () => {
    const { container } = render(<StatsChart count={0} time={0} />);

    expect(container.firstChild).toMatchSnapshot();
  });
});
