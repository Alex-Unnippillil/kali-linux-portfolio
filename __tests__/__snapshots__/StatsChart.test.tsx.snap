// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`StatsChart renders a snapshot for populated values 1`] = `
<svg
  aria-describedby="«r1»"
  aria-labelledby="«r0»"
  class="w-full h-28 mt-2"
  role="img"
  viewBox="0 0 184 120"
>
  <title
    id="«r0»"
  >
    Hash cracking progress overview
  </title>
  <desc
    id="«r1»"
  >
    Bar chart comparing 1,200,000 candidates against 420 seconds.
  </desc>
  <line
    stroke="var(--color-ubt-grey, #aea79f)"
    stroke-width="1.5"
    vector-effect="non-scaling-stroke"
    x1="40"
    x2="40"
    y1="20"
    y2="90"
  />
  <line
    stroke="var(--color-ubt-grey, #aea79f)"
    stroke-width="1.5"
    vector-effect="non-scaling-stroke"
    x1="40"
    x2="164"
    y1="90"
    y2="90"
  />
  <g
    aria-hidden="true"
    fill="var(--color-ubt-grey, #aea79f)"
    font-size="8"
  >
    <text
      text-anchor="end"
      x="36"
      y="96"
    >
      0
    </text>
    <text
      text-anchor="end"
      x="36"
      y="24"
    >
      1,200,000
    </text>
  </g>
  <text
    fill="var(--color-text, #f5f5f5)"
    font-size="9"
    text-anchor="middle"
    transform="rotate(-90 20 55)"
    x="20"
    y="55"
  >
    Value
  </text>
  <text
    fill="var(--color-text, #f5f5f5)"
    font-size="9"
    text-anchor="middle"
    x="102"
    y="108"
  >
    Metrics
  </text>
  <g>
    <rect
      fill="var(--game-color-success)"
      height="70"
      rx="4"
      style="transition: height var(--motion-medium, 300ms) ease-out, y var(--motion-medium, 300ms) ease-out;"
      width="36"
      x="50"
      y="20"
    >
      <title>
        Candidates processed: 1,200,000 candidates
      </title>
    </rect>
    <text
      fill="var(--color-text, #f5f5f5)"
      font-size="8.5"
      text-anchor="middle"
      x="68"
      y="100"
    >
      Candidates
    </text>
  </g>
  <g>
    <rect
      fill="var(--game-color-secondary)"
      height="0.0245"
      rx="4"
      style="transition: height var(--motion-medium, 300ms) ease-out, y var(--motion-medium, 300ms) ease-out;"
      width="36"
      x="118"
      y="89.9755"
    >
      <title>
        Seconds elapsed: 420 seconds
      </title>
    </rect>
    <text
      fill="var(--color-text, #f5f5f5)"
      font-size="8.5"
      text-anchor="middle"
      x="136"
      y="100"
    >
      Seconds
    </text>
  </g>
</svg>
`;

exports[`StatsChart renders a snapshot when values are zero 1`] = `
<svg
  aria-describedby="«r3»"
  aria-labelledby="«r2»"
  class="w-full h-28 mt-2"
  role="img"
  viewBox="0 0 184 120"
>
  <title
    id="«r2»"
  >
    Hash cracking progress overview
  </title>
  <desc
    id="«r3»"
  >
    Bar chart comparing 0 candidates against 0 seconds.
  </desc>
  <line
    stroke="var(--color-ubt-grey, #aea79f)"
    stroke-width="1.5"
    vector-effect="non-scaling-stroke"
    x1="40"
    x2="40"
    y1="20"
    y2="90"
  />
  <line
    stroke="var(--color-ubt-grey, #aea79f)"
    stroke-width="1.5"
    vector-effect="non-scaling-stroke"
    x1="40"
    x2="164"
    y1="90"
    y2="90"
  />
  <g
    aria-hidden="true"
    fill="var(--color-ubt-grey, #aea79f)"
    font-size="8"
  >
    <text
      text-anchor="end"
      x="36"
      y="96"
    >
      0
    </text>
    <text
      text-anchor="end"
      x="36"
      y="24"
    >
      1
    </text>
  </g>
  <text
    fill="var(--color-text, #f5f5f5)"
    font-size="9"
    text-anchor="middle"
    transform="rotate(-90 20 55)"
    x="20"
    y="55"
  >
    Value
  </text>
  <text
    fill="var(--color-text, #f5f5f5)"
    font-size="9"
    text-anchor="middle"
    x="102"
    y="108"
  >
    Metrics
  </text>
  <g>
    <rect
      fill="var(--game-color-success)"
      height="0"
      rx="4"
      style="transition: height var(--motion-medium, 300ms) ease-out, y var(--motion-medium, 300ms) ease-out;"
      width="36"
      x="50"
      y="90"
    >
      <title>
        Candidates processed: 0 candidates
      </title>
    </rect>
    <text
      fill="var(--color-text, #f5f5f5)"
      font-size="8.5"
      text-anchor="middle"
      x="68"
      y="100"
    >
      Candidates
    </text>
  </g>
  <g>
    <rect
      fill="var(--game-color-secondary)"
      height="0"
      rx="4"
      style="transition: height var(--motion-medium, 300ms) ease-out, y var(--motion-medium, 300ms) ease-out;"
      width="36"
      x="118"
      y="90"
    >
      <title>
        Seconds elapsed: 0 seconds
      </title>
    </rect>
    <text
      fill="var(--color-text, #f5f5f5)"
      font-size="8.5"
      text-anchor="middle"
      x="136"
      y="100"
    >
      Seconds
    </text>
  </g>
</svg>
`;
